zebra/ge_netlink.c

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Generic Netlink functions.
+ * Copyright (C) 2022, Carmine Scarpitta
+ */
+
+#include <zebra.h>
+
+#ifdef HAVE_NETLINK
+
+/* The following definition is to workaround an issue in the Linux kernel
+ * header files with redefinition of 'struct in6_addr' in both
+ * netinet/in.h and linux/in6.h.
+ * Reference - https://sourceware.org/ml/libc-alpha/2013-01/msg00599.html
+ */
+#define _LINUX_IN6_H
+
+#include <linux/genetlink.h>
+#include <linux/rtnetlink.h>
+#include <linux/seg6_genl.h>
+
+#include "lib/ns.h"
+#include "zebra/ge_netlink.h"
+#include "zebra/debug.h"
+#include "zebra/kernel_netlink.h"
+
+
+static int16_t seg6_genl_family = -1;
+
+static int genl_parse_getfamily(struct nlmsghdr *h, ns_id_t ns_id, int startup)
```

Comment:
What is exactly seg6_genl_family?  It should be retrieved on startup like all the other startup information.  So FRR doesn't have a special case that it's gathered on first attempt to talk to the data plane.

Reply:
Additionally if retrieval on startup *fails* this can be communicated via the dplane that configuring this will not work properly.  IE -> the main pthread of zebra should be sent the idea of whether or not the generic netlink messaging works if the socket fails to create ( totally possible on older kernels ).  This way when someone enters the encapsulation commands it can be noted this ain't going to work on this kernel.

Reply:
I made the required changes.

`seg6_genl_family` is a numeric identifier for "SRv6" and is used to send "SRv6" commands to the kernel over the Generic Netlink socket.

On startup, zebra queries the kernel to find out if kernel supports configuring SRv6 encap parameters through Generic Netlink and retrieves the `seg6_genl_family` identifier. If it is not supported by the kernel, any future attempt to set SRv6 encap parameters will return an error.